File: my2048.cpp
#include<iostream>
#include<vector>
#include<cstdlib>
#include<iomanip>
using namespace std;

class tile{

public:

    unsigned int value;
    bool blocked;

    tile():value(0),blocked(false){}

};

class game2048{

    vector<vector<tile> >board;
    bool moved,done;
    unsigned int score;


    void addTile(){

        int x,y;
        do{

        x=rand()%4;
        y=rand()%4;
        }while(board[x][y].value);

        rand()%100;
        (rand()%100)>89?board[x][y].value=4:board[x][y].value=2;

        if(canMove())return;

        done=true;

    return;

    }


    bool canMove(){

        for(int i=0;i<4;i++){

            for(int j=0;j<4;j++){

                if(!board[i][j].value)
                        return true;
            }

        }


    for(int i=0;i<4;i++){

        for(int j=0;j<4;j++){

            if(testMove(i+1,j,board[i][j].value))return true;
            if(testMove(i,j+1,board[i][j].value))return true;
            if(testMove(i,j-1,board[i][j].value))return true;
            if(testMove(i-1,j,board[i][j].value))return true;

        }

    }


    return false;
    }


    bool testMove(int i,int j,unsigned int val){

    return (i<0||i>3||j<0||j>3)?false:board[i][j].value==val;


    }




bool is2048Achieved(){

for(int i=0;i<4;i++){
for(int j=0;j<4;j++){
if(board[i][j].value==2048) return true;

}
}
return false;
}

    void transpose(){

        for(int i=0;i<4;i++){
            for(int j=i+1;j<4;j++){

            int temp=board[i][j].value;
            board[i][j].value=board[j][i].value;
            board[j][i].value=temp;
            }

        }

    return;
    }

void displayBoard(){

system("clear");
cout<<setw(38)<<"+----------+"<<endl<<setw(27)<<"|";
cout<<setw(5)<<"  2048!!!"<<setw(2)<<"|"<<endl;
cout<<setw(38)<<"+----------+"<<endl<<endl<<endl;
cout<<"SCORE : "<<score<<endl<<endl;
for(int i=0;i<4;i++){
     cout<<"+-----+-----+-----+-----+"<<endl<<"|";
     for(int j=0;j<4;j++){
     if(!board[i][j].value) cout<<setw(4)<<" ";
     else cout<<setw(4)<<board[i][j].value;
     cout<<" |";
     }
cout<<endl;
}
cout<<"+-----+-----+-----+-----+"<<endl<<endl;
return;
}


        void moveVerti(int i,int j,int d){

   if(board[i][j].value&&!board[i][j].blocked&&!board[i+d][j].blocked&&board[i+d][j].value==board[i][j].value){

                board[i][j].value=0;
                board[i+d][j].value*=2;
                board[i+d][j].blocked=true;
                score+=board[i+d][j].value;
                moved=true;
            }

            else if(!board[i+d][j].value&&board[i][j].value){

                board[i+d][j].value=board[i][j].value;
                board[i][j].value=0;
                moved=true;
            }
        if(d>0)
                    {
          if (i+d<3)
            moveVerti(i+d,j,1);
}
       else
            if(i+d>0)
                moveVerti(i+d,j,-1);


            return;
        }




        void takeInput(){


        moved=false;
        char ch;
        cout<<"UP(W) DOWN(S) LEFT(A) RIGHT(D): "<<endl;
        cin>>ch;
        ch=toupper(ch);
        switch(ch){

            case 'W': for(int j=0;j<4;j++){

                            for(int i=1;i<4;i++){

                                    if(board[i][j].value)moveVerti(i,j,-1);

                                }

                            }

                        break;

            case 'S': for(int j=0;j<4;j++){

                        for(int i=2;i>=0;i--){

                                if(board[i][j].value)moveVerti(i,j,1);

                                }
                            }
                        break;


        case 'A':       transpose();

                        for(int j=0;j<4;j++){

                            for(int i=1;i<4;i++){

                                    if(board[i][j].value)moveVerti(i,j,-1);

                                }

                            }
                            transpose();

                        break;

            case 'D':   transpose();


                    for(int j=0;j<4;j++){

                        for(int i=2;i>=0;i--){

                                if(board[i][j].value)moveVerti(i,j,1);

                                }
                            }
                            transpose();

            break;

};


    for(int i=0;i<4;i++){

        for(int j=0;j<4;j++){

        board[i][j].blocked=false;

        }

    }

    return;
    }

public:

    game2048():board(4,vector<tile>(4)),score(0),moved(true),done(false){}

    void start(){

        addTile();
        while(true){

            if(moved) addTile();
            displayBoard();
            if(is2048Achieved())cout<<"congrats,you made a 2048!!!"<<endl;
            if(done)break;
            takeInput();


        }

    cout<<"Game Over"<<endl;
    return;

    }

};

int main(){


game2048 G;
G.start();

return 0;

}

File: my2048.js
var game=(function(){

var mat=[[0,0,0,0],[0,0,0,0],[0,0,0,0],[0,0,0,0]];
var mat1=[[0,0,0,0],[0,0,0,0],[0,0,0,0],[0,0,0,0]];
var prev_mat=[[0,0,0,0],[0,0,0,0],[0,0,0,0],[0,0,0,0]];
var score,el,flag_undo;

	
	function redraw(){

		console.log("redraw");
		for(var i=0;i<4;i++){
			for(var j=0;j<4;j++){
				console.log(mat[i][j]);
				var x=4*i+j;
				el=document.getElementById(x);
				if(mat[i][j]){
					if(mat[i][j]<=2048){
						var class_name=" _"+mat[i][j];
						el.className="tile";
						el.className+=class_name;
					}
					else{
						el.className="tile_big";
						el.innerHTML=mat[i][j];
					}
				}
				else{
					el.innerHTML="";
					el.className="tile";
				}
			}
		}
		document.getElementById("score").innerHTML="Score= "+score;
	}

	function isGameOver(){
		if(canMove())
			return false;
		
		return true;

	}

	function canMove(){

		for(var i=0;i<4;i++){
			for(var j=0;j<4;j++){
				if(!mat[i][j])
						return true;
			}
		}

		for(var i=0;i<4;i++){
			for(var j=0;j<4;j++){
				if(testMove(i+1,j,mat[i][j])) return true;
				if(testMove(i,j-1,mat[i][j])) return true;
				if(testMove(i,j+1,mat[i][j])) return true;
				if(testMove(i-1,j,mat[i][j])) return true;
			}
		}

	}
	
	function testMove(i,j,value){

		return (i<0||i>3||j<0||j>3)?false:mat[i][j]===value;

	}


	function is2048Achieved(){
		for(var i=0;i<4;i++){
			for(var j=0;j<4;j++){
				if(mat[i][j]===2048)
					return true;
			}
		}
	
		return false;
	}

	function fillRandomEmptyCell(){

		console.log("fill  Random Empty cell")		
		
		do{
			var x=Math.floor(Math.random()*4);
			var y=Math.floor(Math.random()*4);

		}while(mat[x][y]);

		Math.random()<=0.7?mat[x][y]=2:mat[x][y]=4;
		



	}



	function shiftDown(){
		console.log("shift down");
        var i,j;
        for(j=0;j<4;j++){
            var start=0, list_zeroes=[],check=false;
            for(i=3;i>=0;i--){
                if(mat[i][j]===0){
                    check = true;
                    list_zeroes.push(i);
                }
                else if(mat[i][j] && check){
                    mat[list_zeroes[start]][j] = mat[i][j];
                    mat[i][j]=0;
                    list_zeroes.push(i);
                    start++;
                }
            }
        }
	}
	
	function copy(){
		console.log("copy");
		for(var i=0;i<4;i++){
			for(var j=0;j<4;j++){
				mat1[i][j]=mat[i][j];
			}
		}
	}


	function copy_prev_matrix(){

		for(var i=0;i<4;i++){
			for(var j=0;j<4;j++){
				prev_mat[i][j]=mat[i][j];
			}
		}
	}
	function moved(){

		console.log("moved");
		for(var i=0;i<4;i++){
			for(var j=0;j<4;j++){
				if(mat1[i][j]!==mat[i][j])return true;
			}
		}
		return false;
	}

	function moveDown(){

		console.log("move down");
		copy();
				for (var j=0;j<4;j++){
			for (var i=3;i>=0;i--){
				if (mat[i][j]!==0){
					for (var k=i-1;k>=0;k--){
						if (mat[i][j]===mat[k][j]){
							mat[i][j]=2*mat[k][j];
							mat[k][j]=0;
							score+=mat[i][j];
							i=k;
							break;
						}
						else if(mat[k][j]!==0){
							break;
						}
					}
				}
			}
		}
	

		shiftDown();
		if(moved())
				fillRandomEmptyCell();

	}



	function col_reverse(){

		console.log("col_reverse");
		for(var j=0;j<4;j++){
			var temp=mat[0][j];
			mat[0][j]=mat[3][j];
			mat[3][j]=temp;
			temp=mat[1][j];
			mat[1][j]=mat[2][j];
			mat[2][j]=temp;
		}
	}

	function row_reverse(){
		console.log("row_reverse");
		for(var i=0;i<4;i++){
			var temp=mat[i][0];
			mat[i][0]=mat[i][3];
			mat[i][3]=temp;
			temp=mat[i][1];
			mat[i][1]=mat[i][2];
			mat[i][2]=temp;
		}
	}



	function transpose(){
		console.log("transpose");
		for(var i=0;i<4;i++){
			for(var j=i+1;j<4;j++){
				var temp=mat[i][j];
				mat[i][j]=mat[j][i];
				mat[j][i]=temp;
			}
		}
	}


	function move(e){
		e.preventDefault();
		copy_prev_matrix();
		flag_undo=true;
		console.log("something!")
		if(e.keyCode===37){
			console.log("blah");
			transpose();
			col_reverse();// rotate 90 degree anti clockwise
			moveDown();
			transpose();
			row_reverse();
		}

	

		if(e.keyCode===38){

				transpose();
				row_reverse();
				transpose();
				row_reverse();
				moveDown();
				transpose();
				col_reverse();
				transpose();
				col_reverse();
															/// rotate 180 degree clockwise
		}
	

		if(e.keyCode===39){

			transpose();
			row_reverse();
			moveDown();
			transpose();
			col_reverse();								///rotating 90 degree clock wise		

		}

	
		if(e.keyCode===40){

			moveDown();
		}
		redraw();
		if(isGameOver()){
				console.log("game over");
								
				document.getElementById("over").style.height="100%";
	}
		
	} 


	function undo(){

		if(flag_undo){
			for(var i=0;i<4;i++){
				for(var j=0;j<4;j++){
					mat[i][j]=prev_mat[i][j];
				}
			}
			redraw();
			flag_undo=false;		
		}
		else{
			document.getElementById("invalid-undo").style.height="100%";
		}
	}
	function reset(){


		document.getElementById("over").style.height="0%";		
			
		if(localStorage.matrix&&!isGameOver()){
			 mat=JSON.parse(localStorage.matrix);
			
		}
		else{
			mat=[[0,0,0,0],[0,0,0,0],[0,0,0,0],[0,0,0,0]];
			score=0;
			fillRandomEmptyCell();
			fillRandomEmptyCell();
			

		}
	
		redraw();
	}
	
	function BeforeUnload(){

		localStorage.setItem("mat",JSON.stringify(prev_mat));
		


	}
	function init(id){
		
		flag_undo=false;
		reset();
		

		var el1=document.getElementById(id);
		el1.addEventListener("click",reset);
		
		
		document.getElementById("try-again").addEventListener("click",reset);
		document.getElementById("undo").addEventListener("click",undo);
		
		document.getElementById("ok").addEventListener("click",function(){
			document.getElementById("invalid-undo").style.height="0%";
		});
		
		document.body.addEventListener('keydown',move);
		window.addEventListener("unload",BeforeUnload);


}

return {

			init:init
	};

})();